### ISSUE TITLE
Exception thrown by injected MongoOperations are wrapped in undeclared checked InvocationTargetException

### ISSUE THREAD

**Issue by mjustin** [mongock/mongock#514]
## Description

A migration I was involved with involved setting a value of a unique indexed property.  If the value was already in use, the desired outcome was to skip that particular portion of the migration, but still do the rest and report a success.

The initial implementation attempted to catch `org.springframework.dao.DuplicateKeyException`, as that's what `MongoOperations` normally throws.  However, the Mongock injected `MongoOperations` instead throws an `InvocationTargetException` wrapping the `DuplicateKeyException`.

This behavior is problematic for a few reasons:

1. The Mongock `MongoOperations` cannot be used as a direct drop-in replacement for a standard `MongoOperations`, as it behaves differently in this situation (it violates the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)).
2. `InvocationTargetException` is a checked exception type, but `MongoOperations` does not declare that it throws this type.  This violates Java language rules about checked exceptions.  As a consequence, it's impossible to directly catch the `InvocationTargetException`, since Java does not allow catching a checked exception type that is not declared by the method.  Instead, the more general `Exception` type is the most specific exception that can caught to catch this.
3. It makes migration code that cares about the exception type less straightforward, as the exception must be caught and unwrapped before it can be compared and acted upon.

It feels like the `MongoOperations` should throw an exception of the expected type that would be thrown if it weren't wrapped by the Mongock proxy, rather than a language-defying wrapping exception.

## PRIORITY

NORMAL or MINOR

### Mongock
- Mongock version: 5.0.36
- Modules involved: mongock-springboot, mongodb-springdata-v3-driver, MongoDB
- How Mongock is used: annotation approach

### Environment
- mongock-springboot 5.0.36
- mongodb-springdata-v3-driver 5.0.36
- spring-boot-starter-data-mongodb 2.6.4
- MongoDB 4.2.3
- Local MongoDB database

## Steps to Reproduce

1. Create a MongoDB migration that sets a unique value already in use
2. Run the Spring Boot application
3. Observe that `InvocationTargetException` is thrown, rather than `DuplicateKeyException`

### Files

**build.gradle**
```groovy
plugins {
    id 'java'
    id 'org.springframework.boot' version '2.6.4'
}

sourceCompatibility = '17'

repositories {
    mavenCentral()
}

dependencies {
    implementation 'io.mongock:mongock-springboot:5.0.36'
    implementation 'io.mongock:mongodb-springdata-v3-driver:5.0.36'
    implementation 'org.springframework.boot:spring-boot-starter-data-mongodb:2.6.4'
}
```

```java
package com.example.mongock;

import io.mongock.runner.springboot.EnableMongock;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@EnableMongock
@SpringBootApplication
public class Application {
    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }
}
```

**application.yml**
```yaml
mongock.migration-scan-package: com.example.mongock.migrations
```

```java
package com.example.mongock;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;

public class UniquelyIndexed {
    public UniquelyIndexed() {
    }

    public UniquelyIndexed(String uniqueIndex) {
        this.uniqueIndex = uniqueIndex;
    }

    @Id
    public String id;

    @Indexed(unique = true, sparse = true)
    public String uniqueIndex;
}
```

```java
package com.example.mongock.migrations;

import com.example.mongock.UniquelyIndexed;
import io.mongock.api.annotations.ChangeUnit;
import io.mongock.api.annotations.Execution;
import io.mongock.api.annotations.RollbackExecution;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.MongoOperations;
import org.springframework.data.mongodb.core.index.Index;

import java.lang.reflect.InvocationTargetException;

@ChangeUnit(id = "uniqueIndexMigration", order = "1", runAlways = true)
public class UniqueIndexMigration {
    private final MongoOperations mongoOperations;

    public UniqueIndexMigration(MongoOperations mongoOperations) {
        this.mongoOperations = mongoOperations;
    }

    @Execution
    public void migrate() {
        mongoOperations.dropCollection(UniquelyIndexed.class);
        mongoOperations.indexOps(UniquelyIndexed.class).ensureIndex(
                new Index("uniqueIndex", Sort.Direction.ASC).unique().sparse());
        mongoOperations.insert(new UniquelyIndexed("A"));
        try {
            mongoOperations.insert(new UniquelyIndexed("A"));
        } catch (DuplicateKeyException e) {
            // This doesn't occur
            System.out.println("Unique index duplicate key; skipping");
        } catch (Exception e) {
            if (e instanceof InvocationTargetException
                    && e.getCause() instanceof DuplicateKeyException) {
                System.out.println("Encountered DuplicateKeyException "
                        + "wrapped in an InvocationTargetException");
                e.printStackTrace();
            }
        }
        System.out.println("Migration completed");
    }

    @RollbackExecution
    public void rollback() {
        // Not relevant for this example
    }
}
```

### Output:
```
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.mongock.driver.api.lock.guard.proxy.LockGuardProxy.invoke(LockGuardProxy.java:55)
	at io.mongock.driver.api.lock.guard.proxy.LockGuardMethodHandler.invoke(LockGuardMethodHandler.java:22)
	at org.springframework.data.mongodb.core.MongoOperations_$$_mongock_4_0.insert(MongoOperations_$$_mongock_4_0.java)
	at com.example.mongock.migrations.UniqueIndexMigration.migrate(UniqueIndexMigration.java:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.mongock.runner.core.executor.ChangeLogRuntimeImpl.runChangeSet(ChangeLogRuntimeImpl.java:75)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.executeChangeSetMethod(MigrationExecutorBase.java:379)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.executeAndLogChangeSet(MigrationExecutorBase.java:237)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.processSingleChangeSet(MigrationExecutorBase.java:198)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.loopRawChangeSets(MigrationExecutorBase.java:165)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.lambda$processChangeLogInTransactionIfApplies$3(MigrationExecutorBase.java:155)
	at io.mongock.runner.core.executor.NonTransactioner.executeInTransaction(NonTransactioner.java:17)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.processChangeLogInTransactionIfApplies(MigrationExecutorBase.java:155)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.processSingleChangeLog(MigrationExecutorBase.java:129)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.processChangeLogs(MigrationExecutorBase.java:119)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.lambda$processMigration$1(MigrationExecutorBase.java:114)
	at io.mongock.runner.core.executor.NonTransactioner.executeInTransaction(NonTransactioner.java:17)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.processMigration(MigrationExecutorBase.java:114)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.executeMigration(MigrationExecutorBase.java:101)
	at io.mongock.runner.core.executor.operation.change.MigrationExecutorBase.executeMigration(MigrationExecutorBase.java:46)
	at io.mongock.runner.core.executor.MongockRunnerImpl.execute(MongockRunnerImpl.java:53)
	at io.mongock.runner.springboot.base.MongockApplicationRunner.run(MongockApplicationRunner.java:18)
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:768)
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:758)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:310)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1312)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301)
	at com.example.mongock.Application.main(Application.java:11)
Caused by: org.springframework.dao.DuplicateKeyException: Write operation error on server localhost:27017. Write error: WriteError{code=11000, message='E11000 duplicate key error collection: test.uniquelyIndexed index: uniqueIndex_1 dup key: { uniqueIndex: "A" }', details={}}.; nested exception is com.mongodb.MongoWriteException: Write operation error on server localhost:27017. Write error: WriteError{code=11000, message='E11000 duplicate key error collection: test.uniquelyIndexed index: uniqueIndex_1 dup key: { uniqueIndex: "A" }', details={}}.
	at org.springframework.data.mongodb.core.MongoExceptionTranslator.translateExceptionIfPossible(MongoExceptionTranslator.java:106)
	at org.springframework.data.mongodb.core.MongoTemplate.potentiallyConvertRuntimeException(MongoTemplate.java:2906)
	at org.springframework.data.mongodb.core.MongoTemplate.execute(MongoTemplate.java:557)
	at org.springframework.data.mongodb.core.MongoTemplate.insertDocument(MongoTemplate.java:1459)
	at org.springframework.data.mongodb.core.MongoTemplate.doInsert(MongoTemplate.java:1259)
	at org.springframework.data.mongodb.core.MongoTemplate.insert(MongoTemplate.java:1174)
	at org.springframework.data.mongodb.core.MongoTemplate.insert(MongoTemplate.java:1159)
	... 35 more
Caused by: com.mongodb.MongoWriteException: Write operation error on server localhost:27017. Write error: WriteError{code=11000, message='E11000 duplicate key error collection: test.uniquelyIndexed index: uniqueIndex_1 dup key: { uniqueIndex: "A" }', details={}}.
	at com.mongodb.client.internal.MongoCollectionImpl.executeSingleWriteRequest(MongoCollectionImpl.java:1018)
	at com.mongodb.client.internal.MongoCollectionImpl.executeInsertOne(MongoCollectionImpl.java:471)
	at com.mongodb.client.internal.MongoCollectionImpl.insertOne(MongoCollectionImpl.java:454)
	at com.mongodb.client.internal.MongoCollectionImpl.insertOne(MongoCollectionImpl.java:448)
	at org.springframework.data.mongodb.core.MongoTemplate.lambda$insertDocument$16(MongoTemplate.java:1465)
	at org.springframework.data.mongodb.core.MongoTemplate.execute(MongoTemplate.java:555)
	... 39 more
```

## Behaviour

**Expected behavior:**
`MongoOperations` should throw a `DuplicateKeyException`.  It should definitely not throw an undeclared unchecked exception.

**Actual behavior:**
`MongoOperations` throws an undeclared checked `InvocationTargetException` wrapping the `DuplicateKeyException`.

**How often the bug happens:**
Always

**Comment by dieppa** [comment]
we'll take a look and will get back to you as soon as possible

**Comment by osantana85** [comment]
Hello @mjustin , this has been fixed and will be released soon, in version 5.0.38.
